src/main/java/com/mongodb/jdbc/MongoResultSetMetaData.java

```diff
@@ -118,51 +119,68 @@ private void processDataSource(
-        for (NameSpace n : columnIndices) {
-            columnIndexMap.put(Arrays.asList(n.datasource, n.columnLabel), n);
-        }
-
-        // turn columnInfo into a map as well
-        HashMap<List<String>, MongoColumnInfo> columnInfoMap =
-                new HashMap<List<String>, MongoColumnInfo>();
-        for (MongoColumnInfo t : columnInfo) {
-            columnInfoMap.put(Arrays.asList(t.getTableName(), t.getColumnName()), t);
-        }
+    private void processSelectOrder(List<List<String>> selectOrder, MongoJsonSchema schema)
+            throws SQLException {
 
         // reset columnIndices and columnInfo to empty lists and populate in select order
         columnIndices = new ArrayList<NameSpace>();
         columnInfo = new ArrayList<MongoColumnInfo>();
         for (List<String> column : selectOrder) {
-            columnIndices.add(columnIndexMap.remove(column));
-            columnInfo.add(columnInfoMap.remove(column));
+            String datasource = column.get(0);
+            String field = column.get(1);
+            MongoJsonSchema datasourceSchema = schema.properties.get(datasource);
+            assertDatasourceSchema(datasourceSchema);
+            processColumnInfo(datasource, field, datasourceSchema);
+        }
+    }
+
+    private void processColumnInfo(
+            String datasource, String field, MongoJsonSchema datasourceSchema) throws SQLException {
+        MongoJsonSchema columnSchema = datasourceSchema.properties.get(field);
+        BsonTypeInfo columnBsonTypeInfo = columnSchema.getBsonTypeInfo();
+        int nullability = datasourceSchema.getColumnNullability(field);
+        columnIndices.add(new NameSpace(datasource, field));
+        columnInfo.add(new MongoColumnInfo(datasource, field, columnBsonTypeInfo, nullability));
+        if (!columnLabels.containsKey(field)) {
+            List<DatasourceAndIndex> datasourceAndIndexList = new ArrayList<>();
+            datasourceAndIndexList.add(
+                    new DatasourceAndIndex(datasource, columnIndices.size() - 1));
+            columnLabels.put(field, datasourceAndIndexList);
+        } else {
+            columnLabels
+                    .get(field)
+                    .add(new DatasourceAndIndex(datasource, columnIndices.size() - 1));
         }
     }
 
-    // This gets the datasource for a given columnLabel, and is used
-    // in MongoResultSet to retrieve data by label.
-    protected String getDatasource(String columnLabel) {
-        return columnLabels.get(columnLabel).datasource;
+    // This gets the datasource for a given columnLabel.
+    // It is only used in our unit tests.
+    protected String getDatasource(String columnLabel) throws Exception {
+        List<DatasourceAndIndex> columnsForLabel = columnLabels.get(columnLabel);
+        if (columnsForLabel.size() > 1) {
+            throw new Exception(
+                    "Multiple columns with the same label exists. Please use indexes instead.");
```

Comment:
```suggestion
throw new Exception(String.format(
            "Multiple columns with the label '%s' exist. Use indexes to avoid ambiguity.", columnLabel));
```
How about adding the label in the exception?  Also suggesting a change to the rest of the message which you can accept if you prefer it.